File: proj-java/src/main/java/engine/Q1.java
package main.java.engine;


import main.java.common.*;

/**
 * Dado o identificador de um post, a fun¸c˜ao deve retornar
 * o t´ıtulo do post e o nome (n˜ao o ID) de utilizador do autor;
 */
public class Q1 {

    /**
     *
     * @param dados base de dados com toda a informação do sistema
     * @param id id do post a ser procurado
     * @return retorna um par com o titulo da pergunta e o id do user que a fez
     * @throws PostNaoExisteException retorna a excepção de inesistencia do post
     * @throws IdNegativoException retorna erro de introdução de id negativo
     */
    public static Pair<String,String> infoFromPost(Dados dados, long id) throws PostNaoExisteException,IdNegativoException,UserNaoExisteException {
        if(id < 0)  throw new IdNegativoException("Id de Post negativo");
        Pair<String,String> resposta = new Pair<>(null, null);
        Post post;
        post = dados.getPost(id);
        if(post.getTipo() == 1){
            resposta = new Pair<>(post.getTitulo(),dados.getUser(post.getUserId()).getNome());
        }
        if(post.getTipo() == 2){
            post = dados.getPost(post.getParentId());
            resposta = new Pair<>(post.getTitulo(),dados.getUser(post.getUserId()).getNome());
        }
        return resposta;
    }
}

File: proj-c/src/q1.c
#include <string.h>
#include <stdio.h>

#include <arrayList.h>
#include <globalstruct.h>
#include <htable.h>
#include <quickSortDates.h>
#include <real_user.h>
#include <user.h>
#include <interface.h>

/**
@file q1.c
Módulo que contém a funçõe STR_pair info_from_post(TAD_community com, long id).
*/

/**\Função que procura o id de uma pergunta
@param com - estrutura de dados que guarda toda a informação 
@param id - id da pergunta sobre a qual queremos saber o id do user ou
@ resposta da qual queremos saber o id do user que fez a pergunta a qual
@ a resposta corresponde 
@return titulo e id de um user especifico 
*/

STR_pair info_from_post(TAD_community com, long id){ 
	
	STR_pair resposta;
	TAD_POST post = (TAD_POST) getDados(getPosts(com), id);
	if (post == NULL){
	 resposta = create_str_pair(NULL,NULL);
	 return resposta;
	}
	TAD_REAL_USER user = (TAD_REAL_USER) getDados(getUsers(com), getUserId(post));

	//Depois de ter os dados verificamos se se trata de uma pergunta ou uma resposta:
	//Tratando-se de uma pergunta basta-nos obter o nome do utilizador que a fez. 
   	if(getTipo(post) == 1){
		resposta = create_str_pair(getTitulo(post),getName(user)); // post.c-> getpergunta
	}
	
	// Tratando-se de uma resposta identificamos a pergunta correspondente 
	//e de seguida procuramos o nome de utilizador dessa mesma pergunta.
	else{
		int parent = getParentId(post);
		post = getDados(getPosts(com),parent);
		user = getDados(getUsers(com),getUserId(post));
		resposta = create_str_pair(getTitulo(post),getName(user));
	}
    
	return resposta;
}